enable using filesystem from a UNC location

SteveL-MSFT opened the pull request:
The problem is while validating the path, it splits `\\server\share\folder` by the backslash so the root is just `\` and the path ends up being `\server\share\folder` which isn't valid.  Fix is on Windows to add extra slash if the root is `\`

Fix https://github.com/PowerShell/PowerShell/issues/4990

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

SteveL-MSFT committed:
enable using filesystem from a UNC location

anmenaga commented on src/System.Management.Automation/namespaces/NavigationProviderBase.cs:
```diff
@@ -424,6 +424,13 @@ protected string MakePath(string parent, string child, bool childIsLeaf)
+                    // On Windows, if we have a single backslash for the parent, it's a UNC path so we need to add another backslash back
+                    if (String.Compare(parent, StringLiterals.DefaultPathSeparatorString, StringComparison.Ordinal) == 0)
+                    {
+                        builder.Append('\\');
```
Probably better to use DefaultPathSeparatorString instead of char constant.

SteveL-MSFT committed:
[feature]
address PR feedback

SteveL-MSFT commented:
I don't like the change being outside of the filesystemprovider, let me see if I can rework this.

daxian-dbw commented:
@SteveL-MSFT yeah, it would be the best if we can fix the `GetParentPath` to take into account UNC paths.

SteveL-MSFT commented:
@daxian-dbw `GetParentPath` gets called repeatedly to validate the entire path from leaf to root.  I think putting the fix in NormalizePath() may be better.

SteveL-MSFT committed:
[feature]
removed changed in NavigationProviderBase and made change in FileSystemProvider where it should belong

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1:
```diff
@@ -1301,3 +1301,19 @@ Describe "Extended FileSystem Path/Location Cmdlet Provider Tests" -Tags "Featur
 }
+
+Describe "UNC paths" -Tags 'CI' {
+    It "Can Get-ChildItems from a UNC location" -Skip:(!$IsWindows) {
```
Should we check Set-Location too?

SteveL-MSFT replied:
Sure

iSazonov reviewed (commented):
I unlike the fix - it seems a performance is bad.
I do `cd \\localhost\c$` and discover that we call `NormalizePath` 6-9 times for the same path!

SteveL-MSFT committed:
added variations of tests for set-location and push-location
no need to run [feature] anymore since it passed previously and the test case added is CI

SteveL-MSFT commented:
@iSazonov I've noticed that the providers get called a lot.  Perhaps GetParentPath() is the right place even thought that gets called a lot as well, but we only need to make it a UNC path again if it's at the root.

iSazonov commented:
I wonder why we have to call NormalizePath method so many times on the same path? Maybe we can fix this?

SteveL-MSFT committed:
[feature]
move code to reproduce UNC path to GetParentPath()

SteveL-MSFT commented:
@iSazonov I created https://github.com/PowerShell/PowerShell/issues/5002 as the perf work item

adityapatwardhan commented:
@anmenaga  Please update your review.

adityapatwardhan commented:
@anmenaga  ping...

anmenaga reviewed (approved):
LGTM.
Too many paths. :)

adityapatwardhan reviewed (approved):
LGTM
